tensorzero-core/src/db/clickhouse/inference_queries.rs

```diff
@@ -67,48 +65,271 @@ pub(crate) fn generate_list_inferences_sql(
     config: &Config,
     opts: &ListInferencesParams<'_>,
 ) -> Result<(String, Vec<QueryParameter>), Error> {
-    let mut params_map: Vec<QueryParameter> = Vec::new();
-    let mut param_idx_counter = 0; // Counter for unique parameter names
-
-    let function_config = config.get_function(opts.function_name)?;
-    let function_name_param_placeholder = add_parameter(
-        opts.function_name,
-        ClickhouseType::String,
-        &mut params_map,
-        &mut param_idx_counter,
-    );
-    let mut select_clauses = get_select_clauses(&function_config, &function_name_param_placeholder);
-    let mut joins = JoinRegistry::new();
-    let mut where_clauses: Vec<String> = Vec::new();
+    let mut query_params: Vec<QueryParameter> = Vec::new();
+    let mut param_idx_counter = 0;
+
+    let mut sql = match opts.function_name {
+        // If function_name is provided, we know which table to query
+        // TODO(#4181): list_inferences requests with function name should also query both tables
+        // and UNION ALL. The ORDER BY metric join is complicated and net new, so we'll tackle it later.
+        Some(function_name) => {
+            let mut joins = JoinRegistry::new();
+            let function_config = config.get_function(function_name)?;
+            let mut query = generate_single_table_query_for_type(
+                config,
+                opts,
+                function_config.table_name(),
+                function_config.config_type() == FunctionConfigType::Chat,
+                &mut joins,
+                &mut query_params,
+                &mut param_idx_counter,
+            )?;
+
+            // For single-table queries, use proper ORDER BY generation that supports joining with metrics.
+            // Reuse the join registry from the filter so we don't create duplicate joins
+            let order_by_sql = if opts.order_by.is_some() {
+                generate_order_by_sql(
+                    opts.order_by,
+                    config,
+                    &mut query_params,
+                    &mut param_idx_counter,
+                    &mut joins,
+                )?
+            } else {
+                String::new()
+            };
+
+            // Add all joins (from filters and ORDER BY) before WHERE clause
+            if !joins.get_clauses().is_empty() {
+                if let Some(where_pos) = query.find("\nWHERE") {
+                    query.insert_str(where_pos, &joins.get_clauses().join("\n"));
+                } else {
+                    // No WHERE clause, append at end
+                    query.push_str(&joins.get_clauses().join("\n"));
+                }
+            }
```

Comment:
this seems quite brittle; not sure what the solution is though; maybe we should change the return type of generate_single_table_query_for_type to return a more structured object and then render it in this function only.